File: agent/manifest.py
"""扫描 data_dir 下 script_id/version/ 与 version.txt，生成 installed_scripts。"""
from __future__ import annotations

import os

from . import versionutil


def scan_installed_scripts(data_dir: str) -> list[dict]:
    """
    目录结构: data_dir/<script_id>/<version>/version.txt
    返回 [{"script_id","version","env_status"}, ...]
    """
    out: list[dict] = []
    if not os.path.isdir(data_dir):
        return out
    for script_id in os.listdir(data_dir):
        sd = os.path.join(data_dir, script_id)
        if not os.path.isdir(sd):
            continue
        for ver in os.listdir(sd):
            vd = os.path.join(sd, ver)
            vf = os.path.join(vd, "version.txt")
            if not os.path.isfile(vf):
                continue
            try:
                raw = open(vf, encoding="utf-8").read().strip()
            except OSError:
                continue
            nv = versionutil.normalize(raw)
            if not nv:
                continue
            # 简单认为有 version.txt 即 ready（未做 venv/pip 检测）
            out.append(
                {
                    "script_id": script_id,
                    "version": raw,
                    "env_status": "ready",
                }
            )
    return out

File: internal/agentapp/manifest.go
package agentapp

import (
	"os"
	"path/filepath"
	"strings"
	"time"

	v1 "caichip/api/agent/v1"
)

// ScanInstalledScripts 扫描 data_dir，仅上报 **ready** 的脚本（供任务心跳 §2 installed_scripts）。
func ScanInstalledScripts(dataDir string) ([]*v1.InstalledScript, error) {
	rows, err := ScanScriptRowsForSync(dataDir)
	if err != nil {
		return nil, err
	}
	var out []*v1.InstalledScript
	for _, r := range rows {
		if r.GetEnvStatus() != "ready" {
			continue
		}
		out = append(out, &v1.InstalledScript{
			ScriptId:  r.GetScriptId(),
			Version:   r.GetVersion(),
			EnvStatus: "ready",
		})
	}
	return out, nil
}

// ScanScriptRowsForSync 扫描本地脚本包目录，生成 §3 脚本安装心跳的 scripts[]（含 preparing / failed / ready）。
func ScanScriptRowsForSync(dataDir string) ([]*v1.ScriptRow, error) {
	var rows []*v1.ScriptRow
	_ = os.MkdirAll(dataDir, 0755)
	ents, err := os.ReadDir(dataDir)
	if err != nil {
		return nil, err
	}
	for _, e := range ents {
		if !e.IsDir() {
			continue
		}
		scriptID := e.Name()
		sub := filepath.Join(dataDir, scriptID)
		vers, err := os.ReadDir(sub)
		if err != nil {
			continue
		}
		for _, ve := range vers {
			if !ve.IsDir() {
				continue
			}
			vdir := filepath.Join(sub, ve.Name())
			row, ok := scriptRowFromVersionDir(scriptID, ve.Name(), vdir)
			if !ok {
				continue
			}
			rows = append(rows, row)
		}
	}
	return rows, nil
}

func scriptRowFromVersionDir(scriptID, dirName, vdir string) (*v1.ScriptRow, bool) {
	if _, err := os.Stat(filepath.Join(vdir, MarkerPreparing)); err == nil {
		ver := readVersionForRow(vdir, dirName)
		return &v1.ScriptRow{
			ScriptId:  scriptID,
			Version:   ver,
			EnvStatus: "preparing",
		}, true
	}
	if b, err := os.ReadFile(filepath.Join(vdir, MarkerEnvFailed)); err == nil {
		ver := readVersionForRow(vdir, dirName)
		return &v1.ScriptRow{
			ScriptId:  scriptID,
			Version:   ver,
			EnvStatus: "failed",
			Message:   strings.TrimSpace(string(b)),
		}, true
	}
	b, err := os.ReadFile(filepath.Join(vdir, "version.txt"))
	if err != nil {
		return nil, false
	}
	raw := strings.TrimSpace(string(b))
	if raw == "" {
		return nil, false
	}
	row := &v1.ScriptRow{
		ScriptId:  scriptID,
		Version:   raw,
		EnvStatus: "ready",
	}
	if bs, err := os.ReadFile(filepath.Join(vdir, FilePackageSHA256)); err == nil {
		s := strings.TrimSpace(string(bs))
		if s != "" {
			row.PackageSha256 = &s
		}
	}
	if st, err := os.Stat(filepath.Join(vdir, "version.txt")); err == nil {
		t := st.ModTime().UTC().Format(time.RFC3339Nano)
		row.InstalledAt = &t
	}
	return row, true
}

func readVersionForRow(vdir, dirName string) string {
	if b, err := os.ReadFile(filepath.Join(vdir, "version.txt")); err == nil {
		if s := strings.TrimSpace(string(b)); s != "" {
			return s
		}
	}
	return dirName
}
